torch/csrc/distributed/c10d/CUDASymmetricMemory.cu

```diff
@@ -916,15 +917,354 @@ c10::intrusive_ptr<Block> CUDASymmetricMemoryAllocator::find_block(void* ptr) {
   return it->second;
 }
 
-struct RegisterCUDASymmetricMemoryAllocator {
-  RegisterCUDASymmetricMemoryAllocator() {
+// struct RegisterCUDASymmetricMemoryAllocator {
+//   RegisterCUDASymmetricMemoryAllocator() {
+//     register_allocator(
+//         c10::DeviceType::CUDA,
+//         c10::make_intrusive<CUDASymmetricMemoryAllocator>());
+//   }
+// };
+//
+// static RegisterCUDASymmetricMemoryAllocator register_allocator_;
+
+struct NVSHMEMAllocation {
+  void* ptr;
+  size_t buffer_size;
+  int device_idx;
+
+  NVSHMEMAllocation(void* ptr, size_t buffer_size, int device_idx)
+      : ptr(ptr), buffer_size(buffer_size), device_idx(device_idx) {}
+};
+
+class NVSHMEMSymmetricMemory : public SymmetricMemory {
+ public:
+  NVSHMEMSymmetricMemory(
+      std::shared_ptr<NVSHMEMAllocation> allocation,
+      const std::string& group_name)
+      : allocation_(allocation),
```

Comment:
Move the shared pointer into allocation_, have the other members call the methods on "allocation_"